PWGHF/D2H/TableProducer/dataCreatorCharmResoReduced.cxx

```diff
@@ -966,25 +1292,40 @@ struct HfDataCreatorCharmResoReduced {
             selectedTracks[track.globalIndex()] = hfTrackNoParam.lastIndex();
           }
           fillHfCandD = true;
-        } // track loop
-      }
-
+        } // end of loop on bachelor tracks
+      } // end of do tracks
+      // fill D candidate table
       if (fillHfCandD) { // fill candDplus table only once per D candidate, only if at least one V0 is found
-        hfCandD(prongIdsD[0], prongIdsD[1], prongIdsD[2],
-                indexHfReducedCollision,
-                secondaryVertexD[0], secondaryVertexD[1], secondaryVertexD[2],
-                candD.pxProng0(), candD.pyProng0(), candD.pzProng0(),
-                candD.pxProng1(), candD.pyProng1(), candD.pzProng1(),
-                pVecProng2[0], pVecProng2[1], pVecProng2[2],
-                nItsClsDauMin, nTpcCrossRowsDauMin, chi2TpcDauMax, dtype);
-        if constexpr (withMl) {
-          hfCandDMl(bdtScores[0], bdtScores[1], bdtScores[2], -1., -1., -1.);
+        if constexpr (dType == DType::Dstar || dType == DType::Dplus) {
+          hfCandD3Pr(prongIdsD[0], prongIdsD[1], prongIdsD[2],
+                     indexHfReducedCollision,
+                     secondaryVertexD[0], secondaryVertexD[1], secondaryVertexD[2],
+                     candD.pxProng0(), candD.pyProng0(), candD.pzProng0(),
+                     candD.pxProng1(), candD.pyProng1(), candD.pzProng1(),
+                     varUtils.pVectorProng2[0], varUtils.pVectorProng2[1], varUtils.pVectorProng2[2],
+                     nItsClsDauMin, nTpcCrossRowsDauMin, chi2TpcDauMax, dtype);
+          if constexpr (withMl) {
+            hfCandD3PrMl(bdtScores[0], bdtScores[1], bdtScores[2], bdtScores[3], bdtScores[4], bdtScores[5]);
+          }
+        } else if constexpr (dType == DType::D0) {
+          hfCandD2Pr(prongIdsD[0], prongIdsD[1],
+                     indexHfReducedCollision,
+                     secondaryVertexD[0], secondaryVertexD[1], secondaryVertexD[2],
+                     candD.pxProng0(), candD.pyProng0(), candD.pzProng0(),
+                     candD.pxProng1(), candD.pyProng1(), candD.pzProng1(),
+                     nItsClsDauMin, nTpcCrossRowsDauMin, chi2TpcDauMax);
+          if constexpr (withMl) {
```

Comment:
Since you always fill the D0 if it is compatible either with D0 or D0bar or both (which is OK for wrong-sign studies), it would be important to keep track what is the mass hypothesis (to be able to reject candidates in case the D0 was not selected with the right mass hypothesis)